Cython/Build/Tests/TestRecythonize.py

```diff
@@ -1,212 +1,215 @@
-import shutil
 import os
+import shutil
 import tempfile
 import time
 
 import Cython.Build.Dependencies
 import Cython.Utils
 from Cython.TestUtils import CythonTest
 
+SAME = "The result of cytonization is the same"
+INCORRECT = "Incorrect cythonization"
+LINE_BEFORE_IMPLEMENTATION = '  /* "{filename}":{at_line}\n'
+VARS_LINE = '  /*--- Wrapped vars code ---*/\n'
 
-def fresh_cythonize(*args, **kwargs):
-    Cython.Utils.clear_function_caches()
-    Cython.Build.Dependencies._dep_tree = None  # discard method caches
-    Cython.Build.Dependencies.cythonize(*args, **kwargs)
 
 class TestRecythonize(CythonTest):
+    language_level = 3
+    dep_tree = Cython.Build.Dependencies.create_dependency_tree()
+
+    def clear_function_and_Dependencies_caches(self):
+        Cython.Utils.clear_function_caches()
+        Cython.Build.Dependencies._dep_tree = None  # discard method caches
 
     def setUp(self):
         CythonTest.setUp(self)
+        self.clear_function_and_Dependencies_caches()
         self.temp_dir = (
             tempfile.mkdtemp(
                 prefix='recythonize-test',
                 dir='TEST_TMP' if os.path.isdir('TEST_TMP') else None
             )
         )
+        self.src_dir = os.path.join(self.temp_dir, 'src')
+        os.mkdir(self.src_dir)  # self.temp_dir is already a temp
 
     def tearDown(self):
         CythonTest.tearDown(self)
+        self.clear_function_and_Dependencies_caches()
         shutil.rmtree(self.temp_dir)
 
-    def test_recythonize_pyx_on_pxd_change(self):
-
-        src_dir = tempfile.mkdtemp(prefix='src', dir=self.temp_dir)
-
-        a_pxd = os.path.join(src_dir, 'a.pxd')
-        a_pyx = os.path.join(src_dir, 'a.pyx')
-        a_c = os.path.join(src_dir, 'a.c')
-        dep_tree = Cython.Build.Dependencies.create_dependency_tree()
-
-        with open(a_pxd, 'w') as f:
-            f.write('cdef int value\n')
+    def write_to_file(self, path, text):
+        # BUG: without sleep on Linux-like systems,
+        # cythonize not working properly #4245
```

Comment:
```suggestion
        # Make sure that the file has a newer timestamp than any of the previously written files.
```
We probably don't need to do this for all files, though, right? It should be enough to add a sleep whenever we explicitly want a file to be newer.

You could add an additional argument `newer_than=None` and pass the file path that we want to be older. Then we can check the timestamp here and add a `while` loop that runs until `os.touch(path)` results in a newer timestamp.

Reply:
To be clear, my guess is that this works on some systems because creating the files takes sufficiently long there and/or the resolution of file timestamps is sufficiently high on these systems to create a newer timestamp for the second file. Feels like an accident though. We should prevent this from depending on the speed of the system it runs on and make sure that we have a newer timestamp when we depend on it to be newer.

Reply:
> We should prevent this from depending on the speed of the system it runs on and make sure that we have a newer timestamp when we depend on it to be newer.

Exactly! Therefore, #4245 is waiting for its PR, separate from this. In this we have already made a kludge, and it doesn't make much sense to improve it (kludge).

Reply:
> Exactly! Therefore, #4245 is waiting for its PR, separate from this. In this we have already made a kludge, and it doesn't make much sense to improve it (kludge).

Disagree with this. It is perfectly reasonable for Cython to use a timestamp to determine if a file has changed (because there's no other information it can sensibly use, and because the resolution is fine for human use).

The only place that this matters is in this particular test. Therefore there's nothing else to fix except this test. And the correct fix is to wait long enough that the timestamp has changed. There is no broader issue that needs fixing.

Reply:
> The only place that this matters is in this particular test

What if the user encounters such a problem? As for me, if you run cythonization, then it should use the most recent versions of the files, regardless of which system the code is run on and how the system works.

I do not want, even if it is only a handful of users, to face this problem.

Reply:
> > until os.touch(path) results in a newer timestamp
> 
> @scoder `os.utime`? `os.stat`?
> There is no `os.touch`. .. Since python 3.4 we have `pathlib.Path.touch`, but this is not what we want.

It's what I meant, though. Let me rephrase: "add a while loop that runs until an update of the file's timestamp leads to a visibly newer timestamp". Feel free to use `os.stat()` for the "visibly" part. Or `os.path.getmtime()`, preferably.


> What if the user encounters such a problem?

They won't. The failure is the result of an artificial test setup that does not correctly model the situation to test. At least not safely so.

Reply:
Here is the [implementation](https://github.com/cython/cython/pull/4241/commits/a4c1da2a1e9512ec7a110b30bfd8a13a5a136ca6), let's see if it passes the tests

Reply:
Small [fix](https://github.com/cython/cython/pull/4241/commits/d9109485be7c1ddc2200eecb3b764a9803516b5f) for compatibility reasons, but the solution apparently still doesn't work

Reply:
*sigh* I can't debug the solution properly, and there is already an issue dedicated to this, and we already have a way to make this work .. it requires a lot more investigation, the issue is not as obvious as it seemed, maybe we will transfer responsibility for this to the very issue?

One way or another, if you try to solve this problem through me, then the process will be very delayed, it must first of all be someone with a Linux-like system.

Reply:
> 
> 
> Here is the [implementation](https://github.com/cython/cython/pull/4241/commits/a4c1da2a1e9512ec7a110b30bfd8a13a5a136ca6), let's see if it passes the tests

What inside the while loop changes the path's timestamp?

Reply:
> What inside the while loop changes the path's timestamp?
 
@goldenrockefeller  It's not about the while loop, it's just waiting for the timestamp to change.  It is about writing to a file and waiting for the OS to update the timestamp of that file.

Reply:
There is simply an assumption that if the OS has changed the timestamp in the file, the cythonize function can access the new version of the file, but it seems that this is not the case, at least on Linux again it does not work.

Reply:
ohhh, i pushed .. `res` ... ohh that was temporary name ..

Reply:
Even though it doesn't change anything in terms of failures on Linux

Reply:
> 
> 
> > What inside the while loop changes the path's timestamp?
> 
> @goldenrockefeller It's not about the while loop, it's just waiting for the timestamp to change. It is about writing to a file and waiting for the OS to update the timestamp of that file.

Try continuously writing to file until the timestamp changes:
```
        # Make sure the file has a newer timestamp,
        # otherwise cythonize does not work as expected
        # on Linux-like systems #4245
        with open(path, "w") as f:
            f.write(text)
        while os.path.getmtime(path) == timestamp_before_change:
            with open(path, "w") as f:
                f.write(text)
```

Reply:
> Try continuously writing to file until the timestamp changes:

What for? In this case, are the tests run?
(and it looks like a kludge, tbh)

Not sure what you are trying to solve, the timestamp is changing, but the problem persists, it's not about the timestamp.
Take a look at the test output in the case of Linux. [example](https://github.com/cython/cython/pull/4241/checks?check_run_id=2955886458#step:5:14137)

Reply:
On Linux, some process takes relatively long time, so `cythonize` doesn't see the latest version of files, so the easiest way to deal with this in this PR is to just [wait a bit](https://github.com/cython/cython/pull/4241#issuecomment-867390141)

Regardless of the source of the problem and the possibility of solving it inside `cythonize`, I think this is what needs to be solved in #4245, and not try to guess the reason for it in this PR. It just delays the merge, thus adding useful testing.

Reply:
> the possibility of solving it inside `cythonize`

As I said earlier, since `all_dependencies` always detects a newer version of files (because it never crashes), I think it can be solved by changing something inside `cythonize`, but that requires a lot more research. This is problematic for me, since I work under Windows and do not have this problem.

Reply:
If someone has an intention to solve this problem, they would create a separate PR related to the #4245 and there they would already try to find a good solution.

Reply:
> Not sure what you are trying to solve, the timestamp is changing, but the problem persists, it's not about the timestamp.

Waiting before the write helps, but waiting after the write doesn't. I am thinking that trying out alternative solutions like putting `time.sleep(1)` after the writing to file or doing what I suggested will help investigate the problem. If this is really about timestamp resolution on Ubuntu, or how Ubuntu processes are scheduled, there is not much I can doing on the cython side to fix this problem. Maybe, force a recythonization if the current time is within a few seconds of the file's last modification time? (whether or not the file actually changed of not, since we can not know for sure)

Reply:
Good assumptions @goldenrockefeller, however, I would rather suggest that you test them not through me, but by creating a [new PR based on this](https://github.com/cython/cython/pull/4241#issuecomment-866215524) (I can tell you what is needed for github desktop, I sort of figured out how this can be done) It will be faster this way

.. and I'm already tired of this desire to solve everything on the spot no matter what.

Reply:
Yes .. the conversation went a little too far, and went beyond the original topic of the conversation, I will resolve this conversation.

Reply:
And scoder has already written the solution himself